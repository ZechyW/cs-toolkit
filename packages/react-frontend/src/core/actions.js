/**
 * Core/global actions
 */
import { createAction } from "redux-starter-kit";

// Resets
export const resetAll = createAction("reset/all");
