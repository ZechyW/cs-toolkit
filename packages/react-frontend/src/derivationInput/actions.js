import axios from "axios";
import { createAction } from "redux-starter-kit";

// Lexical item input
export const addItem = createAction("derivationInput/addItem");
export const deleteItemAtIndex = createAction(
  "derivationInput/deleteItemAtIndex"
);
export const changeItemIndex = createAction("derivationInput/changeItemIndex");

// -'~'-.,__,.-'~'-.,__,.-'~'-.,__,.-'~'-.,__,.-'~'-.,__,.-'~'-.,__,.-'~'-.,_
// Submit derivation generation request
const POST_DERIVATION_REQUEST = "derivationInput/postRequest";
const POST_DERIVATION_REQUEST_LOADING = "derivationInput/postRequestLoading";
const POST_DERIVATION_REQUEST_SUCCESS = "derivationInput/postRequestSuccess";
const POST_DERIVATION_REQUEST_ERROR = "derivationInput/postRequestError";

export const postDerivationRequestLoading = createAction(
  POST_DERIVATION_REQUEST_LOADING
);
export const postDerivationRequestSuccess = createAction(
  POST_DERIVATION_REQUEST_SUCCESS
);
export const postDerivationRequestError = createAction(
  POST_DERIVATION_REQUEST_ERROR
);

/**
 * Thunk to post a derivation request to the backend.
 * @return {Function}
 */
export function postDerivationRequest(payload) {
  return async (dispatch) => {
    dispatch(postDerivationRequestLoading(true));

    try {
      const response = await axios.post("/api/grammar/", {
        derivationInput: payload
      });

      if (response.status === 200) {
        dispatch(postDerivationRequestSuccess(response.data));
      } else {
        dispatch(postDerivationRequestError(response));
      }
    } catch (err) {
      dispatch(postDerivationRequestError(err));
    }
  };
}
postDerivationRequest.toString = () => POST_DERIVATION_REQUEST;
postDerivationRequest.type = POST_DERIVATION_REQUEST;
