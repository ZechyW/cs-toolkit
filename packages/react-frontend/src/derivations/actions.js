import { createAction } from "redux-starter-kit";

export const reset = createAction("derivations/reset");

// -'~'-.,__,.-'~'-.,__,.-'~'-.,__,.-'~'-.,__,.-'~'-.,__,.-'~'-.,__,.-'~'-.,_
// `ag-grid` state
export const saveColumnState = createAction("derivations/saveColumnState");
