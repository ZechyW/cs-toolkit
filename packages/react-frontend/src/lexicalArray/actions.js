import { createAction } from "redux-starter-kit";

export const replaceSuggestions = createAction(
  "lexicalArray/replaceSuggestions"
);
export const addItem = createAction("lexicalArray/addItem");
export const deleteItemAtIndex = createAction("lexicalArray/deleteItemAtIndex");
