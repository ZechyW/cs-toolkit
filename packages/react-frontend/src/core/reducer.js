import { createReducer } from "redux-starter-kit";

import { actions as coreActions } from "../core";
import { hideItem, showItem } from "./actions";

const initialState = {
  itemVisibility: {
    derivationInput: true,
    lexicalItemList: true,
    derivationStatusList: true,
    derivationViewer: true
  }
};

export default createReducer(initialState, {
  [showItem]: (state, action) => {
    state.itemVisibility[action.payload] = true;
  },
  [hideItem]: (state, action) => {
    state.itemVisibility[action.payload] = false;
  },

  // Resetting the grid should also reset item visibility options.
  [coreActions.resetGrid]: () => {
    return initialState;
  }
});
