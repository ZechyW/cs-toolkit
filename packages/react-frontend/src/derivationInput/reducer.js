import { createReducer } from "redux-starter-kit";
import { addItem, changeItemIndex, deleteItemAtIndex } from "./actions";

const derivationInputReducer = createReducer(
  {
    currentInput: []
  },
  {
    // Input items
    [addItem]: (state, action) => {
      const { item } = action.payload;
      state.currentInput.push(item);
    },
    [deleteItemAtIndex]: (state, action) => {
      const { index } = action.payload;
      state.currentInput = state.currentInput.filter((item, i) => index !== i);
    },
    [changeItemIndex]: (state, action) => {
      const { item, oldIndex, newIndex } = action.payload;
      state.currentInput.splice(oldIndex, 1);
      state.currentInput.splice(newIndex, 0, item);
    }
  }
);
export default derivationInputReducer;
