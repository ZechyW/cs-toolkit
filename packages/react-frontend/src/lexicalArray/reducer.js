import { uniqBy } from "lodash-es";
import { createReducer } from "redux-starter-kit";
import { addItem, deleteItemAtIndex, replaceSuggestions } from "./actions";

const lexicalArrayReducer = createReducer(
  {
    suggestions: [],
    currentInput: []
  },
  {
    // Suggestions
    [replaceSuggestions]: replaceSuggestionsReducer,

    // Input items
    [addItem]: (state, action) => {
      state.currentInput.push(action.payload);
    },
    [deleteItemAtIndex]: (state, action) => {
      const index = action.payload;
      state.currentInput = state.currentInput.filter((item, i) => index !== i);
    }
  }
);

/**
 * Sets the list of lexical item suggestions based on the provided payload.
 *
 * The list of lexical items may contain duplicates (for items with the
 * same text/language but different features), so we need to deduplicate
 * as we go.
 */
export function replaceSuggestionsReducer(state, action) {
  const newSuggestions = [];

  // Deduplicate by suggestion label
  const lexicalItems = uniqBy(
    action.payload,
    (item) => `${item.text} (${item.language})`
  );

  for (const lexicalItem of lexicalItems) {
    // Suggestions need to have an `id` field and `label` field, and we
    // also track the `text` and `language` fields of lexical items directly.
    newSuggestions.push({
      id: `${lexicalItem.text}`,
      text: `${lexicalItem.text}`,
      label: `${lexicalItem.text} (${lexicalItem.language})`,
      language: lexicalItem.language
    });
  }

  state.suggestions = newSuggestions;
}

export default lexicalArrayReducer;
