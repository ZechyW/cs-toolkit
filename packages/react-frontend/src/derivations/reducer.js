import { createReducer } from "redux-starter-kit";
import { actions as derivationInputActions } from "../derivationInput";
import { actions as wsActions } from "../websocket";
import { reset, saveColumnState } from "./actions";

const initialState = {
  requestsById: {},
  derivationsById: {},

  // ag-grid view
  columnState: []
};

export default createReducer(initialState, {
  [reset]: () => {
    return initialState;
  },

  /**
   * The UI wants to save the current `ag-grid` column state.
   * @param state
   * @param action
   */
  [saveColumnState]: (state, action) => {
    state.columnState = action.payload;
  },

  /**
   * A DerivationRequest was successfully posted to the backend.
   * @param state
   * @param action
   */
  [derivationInputActions.postDerivationRequestSuccess]: (state, action) => {
    const request = action.payload;
    state.requestsById[request.id] = request;
  },

  /**
   * Successfully subscribed to a particular Derivation/DerivationRequest.
   * @param state
   * @param action
   */
  [wsActions.subscribeAcknowledge]: (state, action) => {
    if (action.payload.model === "Derivation") {
      const derivation = action.payload.data;
      state.derivationsById[derivation.id] = derivation;
    }

    if (action.payload.model === "DerivationRequest") {
      const request = action.payload.data;
      state.requestsById[request.id] = request;
    }
  },

  /**
   * A Derivation/DerivationRequest was changed/added.
   * @param state
   * @param action
   */
  [wsActions.subscribeChange]: (state, action) => {
    if (action.payload.model === "Derivation") {
      const derivation = action.payload.data;
      state.derivationsById[derivation.id] = derivation;
    }

    if (action.payload.model === "DerivationRequest") {
      const request = action.payload.data;
      state.requestsById[request.id] = request;
    }
  },

  /**
   * A Derivation/DerivationRequest was deleted.
   * @param state
   * @param action
   */
  [wsActions.subscribeDelete]: (state, action) => {
    if (action.payload.model === "Derivation") {
      const derivation = action.payload.data;
      delete state.derivationsById[derivation.id];
    }

    if (action.payload.model === "DerivationRequest") {
      const request = action.payload.data;
      delete state.requestsById[request.id];
    }
  },

  /**
   * There was a problem subscribing to the given Derivation/DerivationRequest.
   * @param state
   * @param action
   */
  [wsActions.subscribeError]: (state, action) => {
    // Stop tracking any Derivations/DerivationRequests that are no longer
    // valid.  (This could happen if the database is cleared out from the
    // server end)
    if (action.payload.model === "Derivation") {
      delete state.derivationsById[action.payload.id];
    }

    if (action.payload.model === "DerivationRequest") {
      delete state.requestsById[action.payload.id];
    }
  }
});
