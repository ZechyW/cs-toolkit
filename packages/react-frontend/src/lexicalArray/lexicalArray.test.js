import { replaceSuggestions } from "./actions";
import reducer from "./reducer";

describe("The reducer", () => {
  it("returns the initial state", () => {
    expect(reducer(undefined, {})).toEqual({
      suggestions: [],
      currentInput: []
    });
  });

  it("deduplicates suggestions", () => {
    expect(
      reducer(
        { suggestions: [] },
        replaceSuggestions([
          { text: "word", language: "en" },
          { text: "word", language: "en_SG" },
          { text: "word", language: "en" }
        ])
      )
    ).toEqual({
      suggestions: [
        { id: "word", label: "word (en)", language: "en", text: "word" },
        { id: "word", label: "word (en_SG)", language: "en_SG", text: "word" }
      ]
    });
  });
});
