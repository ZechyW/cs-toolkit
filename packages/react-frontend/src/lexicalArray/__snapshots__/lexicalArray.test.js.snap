// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The reducer returns the initial state 1`] = `
Object {
  "currentInput": Array [],
  "suggestions": Array [],
}
`;
